OpenRA.Mods.Common/Warheads/HealthPercentageDamageWarhead.cs

```diff
@@ -43,12 +43,20 @@ public override void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> dama
 			if (!IsValidAgainst(victim, firedBy))
 				return;
 
+			// Cannot be damaged without a Health trait
 			var healthInfo = victim.Info.TraitInfoOrDefault<HealthInfo>();
 			if (healthInfo == null)
 				return;
 
+			// Cannot be damaged without an active HitShape
+			var activeShapes = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled);
+			if (!activeShapes.Any())
+				return;
+
+			var closestActiveShape = activeShapes.MinBy(t => t.Info.Type.DistanceFromEdge(victim.CenterPosition, victim));
```

Comment:
One pass option:
```c#
var closestActiveShape = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled)
    .MinByOrDefault(t => t.Info.Type.DistanceFromEdge(victim.CenterPosition, victim));
if (closestActiveShape == null)
    return;
```
Requires you to allocate the MinBy lambda even if there are no active hit shapes, but I'm assuming that's the uncommon case/a worthwhile tradeoff.